processing/src/main/java/org/apache/druid/query/context/ResponseContext.java

```diff
@@ -76,56 +253,158 @@ public static ResponseContext createEmpty()
+   * @throws IllegalStateException if the key has not been registered.
+   */
+  public Object add(BaseKey key, Object value)
   {
-    getDelegate().putAll(m);
+    final BaseKey registeredKey = Key.keyOf(key.getName());
+    return getDelegate().merge(registeredKey, value, key.getMergeFunction());
   }
 
-  public void putAll(ResponseContext responseContext)
+  /**
+   * Merges a response context into the current.
+   * @throws IllegalStateException If a key of the {@code responseContext} has not been registered.
+   */
+  public void merge(ResponseContext responseContext)
   {
-    getDelegate().putAll(responseContext.getDelegate());
+    responseContext.getDelegate().forEach((key, newValue) -> {
+      if (newValue != null) {
+        add(key, newValue);
+      }
+    });
   }
 
-  public int size()
+  /**
+   * Serializes the context given that the resulting string length is less than the provided limit.
+   * This method tries to remove some elements from context collections if it's needed to satisfy the limit.
+   * The resulting string might be correctly deserialized to {@link ResponseContext}.
+   */
+  public SerializationResult serializeWith(ObjectMapper objectMapper, int maxCharsNumber) throws JsonProcessingException
   {
-    return getDelegate().size();
+    final String fullSerializedString = objectMapper.writeValueAsString(getDelegate());
+    if (fullSerializedString.length() <= maxCharsNumber) {
+      return new SerializationResult(fullSerializedString, fullSerializedString);
+    } else {
+      // Indicates that the context is truncated during serialization.
+      add(Key.TRUNCATED, true);
+      final ObjectNode contextJsonNode = objectMapper.valueToTree(getDelegate());
+      final ArrayList<Map.Entry<String, JsonNode>> sortedNodesByLength = Lists.newArrayList(contextJsonNode.fields());
+      final Comparator<Map.Entry<String, JsonNode>> valueLengthReversedComparator =
+          Comparator.comparing((Map.Entry<String, JsonNode> e) -> e.getValue().toString().length()).reversed();
+      sortedNodesByLength.sort(valueLengthReversedComparator);
+      int needToRemoveCharsNumber = fullSerializedString.length() - maxCharsNumber;
+      // The complexity of this block is O(n*m*log(m)) where n - context size, m - context's array size
+      for (Map.Entry<String, JsonNode> e : sortedNodesByLength) {
+        final String fieldName = e.getKey();
+        final JsonNode node = e.getValue();
+        if (node.isArray()) {
+          if (needToRemoveCharsNumber >= node.toString().length()) {
+            final int lengthBeforeRemove = node.toString().length();
+            // Empty array could be correctly deserialized so we remove only its elements.
+            ((ArrayNode) node).removeAll();
+            final int lengthAfterRemove = node.toString().length();
+            needToRemoveCharsNumber -= lengthBeforeRemove - lengthAfterRemove;
+          } else {
+            final ArrayNode arrNode = (ArrayNode) node;
```

Comment:
This block needs a comment. It's not obvious what and why is going on here. Please extract as a method (or the upper block) if possible.

Reply:
added a comment and extracted